relayer/src/channel/version.rs

```diff
@@ -0,0 +1,135 @@
+            info!("resolving channel version for port id='{}', context='{:?}' by retrieving dst chain '{}' app version", port_id, ctx, dst_chain_id);
+
+            // Note the compatibility logic below:
+            // The destination chain may or may not implement `QueryAppVersionRequest`,
+            // which is only available from ibc-go v2.0.0 or newer versions.
+            // We detect if it is implemented by calling `is_unimplemented_port_query`.
+            // If unimplemented, we can recover by simply using the default version.
+            // This ensures compatibility with ibc-go v2.0.0+ and pre-v2.0.0.
+            match dst_app_version(channel) {
+                Ok(v) => Ok(v),
+                Err(e) => {
+                    if let ErrorDetail::GrpcStatus(s) = e.detail() {
+                        if s.is_unimplemented_port_query() {
+                            // Ensure compatibility & recover from the error,
+                            // by using the default version.
+                            debug!("resorting to default version because dst chain '{}' does not expose app version gRPC endpoint", dst_chain_id);
+                            return default_by_port(port_id);
+                        }
+                    }
+                    // Propagate the error, this is not a recoverable problem.
+                    Err(e).map_err(|e| ChannelError::query_app_version(channel.dst_chain().id(), e))
+                }
+            }
+        }
+    }
+}
+
+/// Returns the default channel version, depending on the
+/// the given [`PortId`].
+pub fn default_by_port(port_id: &PortId) -> Result<Version, ChannelError> {
+    if port_id.as_str() == ics20_fungible_token_transfer::PORT_ID {
+        // https://github.com/cosmos/ibc/tree/master/spec/app/ics-020-fungible-token-transfer#forwards-compatibility
+        Ok(Version::ics20())
+    } else if port_id
+        .as_str()
+        .starts_with(ics27_interchain_accounts::PORT_ID_PREFIX)
+    {
+        // https://github.com/cosmos/ibc/tree/master/spec/app/ics-027-interchain-accounts#channel-lifecycle-management
+        Ok(Version::ics27())
+    } else {
+        warn!(
+            "cannot resolve channel version for unknown port id '{}'",
+            port_id,
+        );
+        Err(ChannelError::invalid_port_id(port_id.clone()))
+    }
+}
+
+/// Fetches the application version from the destination
+/// chain for the given [`Channel`].
+fn dst_app_version<ChainA: ChainHandle, ChainB: ChainHandle>(
+    channel: &Channel<ChainA, ChainB>,
+) -> Result<Version, Error> {
+    let counterparty = channel.src_channel_id().map(|cid| Counterparty {
+        port_id: channel.src_port_id().to_string(),
+        channel_id: cid.to_string(),
+    });
+
+    let proposed_version = channel
+        .src_version()
+        .cloned()
+        .unwrap_or(
+            default_by_port(channel.src_port_id())
+                .map_err(|e| Error::app_version(e.to_string()))?,
+        )
```

Comment:
```suggestion
        .unwrap_or_else(|| {
            default_by_port(channel.src_port_id())
                .map_err(|e| Error::app_version(e.to_string()))?
        })
```

Reply:
I tried following the suggestion but something's off.

```
error[E0277]: the `?` operator can only be used in a closure that returns `Result` or `Option` (or another type that implements `std::ops::FromResidual`)
   --> relayer/src/channel/version.rs:116:98
    |
115 |           .unwrap_or_else(|| {
    |  _________________________-
116 | |             default_by_port(channel.src_port_id()).map_err(|e| Error::app_version(e.to_string()))?
    | |                                                                                                  ^ cannot use the `?` operator in a closure that returns `ibc::core::ics04_channel::Version`
117 | |         })
    | |_________- this function should return `Result` or `Option` to accept `?`
    |
    = help: the trait `std::ops::FromResidual<std::result::Result<std::convert::Infallible, error::Error>>` is not implemented for `ibc::core::ics04_channel::Version`
note: required by `std::ops::FromResidual::from_residual`
   --> /Users/adi/.rustup/toolchains/1.56-x86_64-apple-darwin/lib/rustlib/src/rust/library/core/src/ops/try_trait.rs:339:5
    |
339 |     fn from_residual(residual: R) -> Self;
    |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

```

Any further thoughts?

Reply:
https://github.com/informalsystems/ibc-rs/pull/1582/commits/f25cff663b07ee35d2c85c5b68f07baa1e450415